[MetaxGPU] Support FastDeploy on metax gpu

Kane2011 opened the pull request.

Kane2011 committed:
[MetaxGPU] Support FastDeploy on metax gpu

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PaddlePaddle/FastDeploy?pullRequest=3241) <br/>All committers have signed the CLA.

Kane2011 committed:
Merge branch 'develop' into pr0807

Kane2011 committed:
Merge branch 'develop' into pr0807

Kane2011 committed:
Merge branch 'develop' into pr0807

yongqiangma committed:
Merge branch 'develop' into pr0807

yongqiangma reviewed (dismissed):
LGTM

gongshaotian commented on fastdeploy/worker/metax_worker.py:
```diff
@@ -0,0 +1,216 @@
+from fastdeploy.worker.output import ModelRunnerOutput
+from fastdeploy.worker.worker_base import WorkerBase
+
+logger = get_logger("gpu_worker", "gpu_worker.log")
```
log 命名没有修改

Kane2011 replied:
收到，review完毕我一起处理下， 感谢

gongshaotian commented on fastdeploy/worker/metax_worker.py:
```diff
@@ -0,0 +1,216 @@
+            gc.collect()
+            paddle.device.cuda.empty_cache()
+            if self.parallel_config.enable_custom_all_reduce:
+                from fastdeploy.distributed.communication import use_custom_allreduce
```
custom_allreduce() 算子有适配吗

Kane2011 replied:
目前还没有。

yuanlehome commented on fastdeploy/model_executor/layers/backends/metax/moe/__init__.py:
```diff
@@ -0,0 +1,20 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" "
+metax gpu moe
+"""
```
这三行delete掉哈

Kane2011 replied:
OK

Kane2011 replied:
OK

yuanlehome commented on fastdeploy/model_executor/layers/backends/metax/moe/__init__.py:
```diff
@@ -0,0 +1,20 @@
+metax gpu moe
+"""
+
+__all__ = [
+    "fused_moe_kernel_paddle",
+]
```
没有import fused_moe_kernel_paddle，这个__all__并没有work吧？

Kane2011 replied:
目前没有使用， 可以去掉。

yuanlehome commented on fastdeploy/model_executor/layers/backends/metax/__init__.py:
```diff
@@ -0,0 +1,25 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+metax gpu backend methods
+"""
```
delete it

Kane2011 replied:
OK

gongshaotian commented on fastdeploy/worker/metax_worker.py:
```diff
@@ -0,0 +1,216 @@
+        if self.model_runner.graph_opt_level >= 1:
+            self.model_runner.sot_warmup()
+        # Triger cuda grpah capture
+        self.model_runner.capture_model()
```
Metax 平台支持 CudaGraph 功能吗

Kane2011 replied:
我们有graph功能， 但是目前没有enable 这个功能做调试。

gongshaotian replied:
> 我们有graph功能， 但是目前没有enable 这个功能做调试。

好的，GraphOptBackend 当前没预留多硬件接入接口，近期会支持

gongshaotian commented:
由于合入的代码较多，Worker->ModelRunner 中有许多直接从GPU中迁移过来功能代码，不太好分辨哪些是已经对齐了GPU，哪些是还在支持（例如 CudaGraph、 Custom All Reduce）。能否完善下PR描述，列举下当前 MetaxGpu 能够支持的功能，以及未支持功能的开发进展和计划 🚀

Kane2011 commented:
> 由于合入的代码较多，Worker->ModelRunner 中有许多直接从GPU中迁移过来功能代码，不太好分辨哪些是已经对齐了GPU，哪些是还在支持（例如 CudaGraph、 Custom All Reduce）。能否完善下PR描述，列举下当前 MetaxGpu 能够支持的功能，以及未支持功能的开发进展和计划 🚀

> MX：好的，针对review 的comment，我整理后更新一下。

Kane2011 committed:
Update metax_worker.py

1. change worker log;
2. remove custom allreduce, adapt it later;
3. remove cuda graph;

Kane2011 committed:
Update __init__.py

1. remove metax's key work comment

Kane2011 committed:
Update __init__.py

1. remove metax's key word comment;
2. add fused_moe_kernel_paddle import

yuanlehome commented on fastdeploy/worker/metax_model_runner.py:
```diff
@@ -0,0 +1,1664 @@
+                self.share_inputs["is_block_step"][idx : idx + 1] = False
+                continue
+
+            if len(request.eos_token_ids) < self.parallel_config.eos_tokens_lens:
+                request.eos_token_ids.append(request.eos_token_ids[0])
+            self.share_inputs["eos_token_id"][:] = np.array(request.eos_token_ids, dtype="int64").reshape(-1, 1)
```
这一块请参考此pr修改一下 https://github.com/PaddlePaddle/FastDeploy/pull/3274

yuanlehome replied:
建议merge一下develop再验证下，pad_token_id等不兼容修改，应该会报错

Kane2011 replied:
好的， 可以把这笔先合入后，我再适配这个改动?

yuanlehome reviewed (approved):
LGTM
